src/or/onion.c

```diff
@@ -1339,3 +1342,466 @@ extended_cell_format(uint8_t *command_out, uint16_t *len_out,
+                   const uint8_t *padding_data,
+                   const uint16_t padding_len)
+{
+  tor_assert(cell_out);
+
+  check_handshake_and_padding_len(handshake_type, handshake_len,
+                                  padding_len, true);
+
+  create2v_cell_body_set_htype(cell_out->body, handshake_type);
+  create2v_cell_body_set_hlen(cell_out->body, handshake_len);
+  create2v_cell_body_setlen_hdata(cell_out->body, (size_t)handshake_len);
+  create2v_cell_body_setlen_ignored(cell_out->body, (size_t)padding_len);
+
+  if (handshake_len)
+    memcpy(create2v_cell_body_getarray_hdata(cell_out->body),
+           handshake_data, handshake_len);
+  if (padding_len)
+    memcpy(create2v_cell_body_getarray_ignored(cell_out->body),
+           padding_data, padding_len);
+}
+
+/**
+ * Deallocate a CREATE2V cell.
+ */
+void
+create2v_cell_free(create2v_cell_t *cell)
+{
+  create2v_cell_body_free(cell->body);
+  tor_free(cell);
+}
+
+/**
+ * Given the Create2VMaximumData consensus parameter, calculate the
+ * total maximum allowable size of an incoming CREATE(D)2V buffer.
+ */
+static int
+create2v_cell_calculate_maximum_total(void)
+{
+  return networkstatus_get_create2v_maximum_data(NULL)
+    // XXXisis what about these extra bits of stuff? are we only counting
+    // "ignored" and "hdata" as data? or should we also count all of this?
+    + 2 // for the uint16_t htype
+    + 2 // for the uint16_t hlen
+    + sizeof(size_t) * 4 // for n_ and allocated_ in TRUNNEL_DYNARRAY_HEADs
+    + 1; // for the uint8_t trunnel_error_code_
+}
+
+/**
+ * Check that the encoded length of an incoming CREATE2V or CREATED2V cell is
+ * less than the maximum allowed.
+ *
+ * Returns true if it's okay, and false otherwise.
+ */
+static bool
+create2v_cell_check_encoded_length(const size_t len)
+{
+  const uint16_t max_len = create2v_cell_calculate_maximum_total();
+
+  /* If the incoming buffer is reportedly way larger than that which would fill
+   * the allowable space for hdata and the ignored padding, then fail parsing
+   * early.
+   */
+  if (len > max_len) {
+    log_info(LD_OR, "Received a suspicious CREATE(D)2V cell whose reported "
+             "length (%zd bytes) was greater that the maximum allowed (%hd "
+             "bytes). This is either an attempted DoS, or there exist "
+             "newer/other onion routers speaking a circuit handshake we "
+             "don't know about.", len, max_len);
+    return false;
+  }
+
+  if (len == 0) {
+    log_info(LD_OR, "Received a strange CREATE(D)2V cell whose reported "
+             "length was 0 bytes.");
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Parse the CREATE2V payload at <b>p</b>, which could be up to <b>p_len</b>
+ * bytes long, and use it to fill the fields of <b>cell_out->body</b>.
+ *
+ * Note that part of the body of an EXTEND2V cell is a CREATE2V payload, so
+ * this function is also used for parsing those.
+ *
+ * Returns true if parsing was successful, and false otherwise.
```

Comment:
Hmmmmm.  This is very different than the 0 on success, negative on failure convention we use most everywhere else.  Are we sure we want to start using bools like this?  If we do, it will be hard to read code that checks a function's output for success.....